File: src/ScalaSolutions/scalasolutions/src/easy/TwoSum.scala
package easy

def twoSum(nums: Array[Int], y: Int): Array[Int] = {
  def twoSumR(xs: List[(Int, Int)]): Array[Int] = xs match {
    case Nil => Array.empty[Int]
    case x1 :: xsTail =>
      val answer = twoSumH(y - x1._1, xsTail)
      if (answer.isEmpty) twoSumR(xsTail)
      else Array(x1._2, answer.head._2)
  }

  def twoSumH(target: Int, xs: List[(Int, Int)]): List[(Int, Int)] =
    xs match {
      case Nil => Nil
      case x2 :: xsTail =>
        if (x2._1 == target) x2 :: Nil
        else twoSumH(target, xsTail)
    }

  twoSumR(nums.zipWithIndex.toList)
}

object TwoSum extends App {
  println(twoSum(Array(1, 2, 3, 4, 10, 11, 12, 2, 3), 5).toList)
}

File: src/HaskellSolutions/src/Easy/TwoSum.hs
module Easy.TwoSum where

import           Data.Bool (bool)

twoSum :: [Int] -> Int -> [Int]
twoSum nums y = twoSumR (zip nums [0 ..])
  where
    twoSumR :: [(Int, Int)] -> [Int]
    twoSumR [] = []
    twoSumR (x1:xs) = bool answer (twoSumR xs) (null answer)
      where
        answer = twoSumH (fst x1 +) xs
        twoSumH f []      = []
        twoSumH f (x2:xs) = bool (twoSumH f xs) [snd x1, snd x2] (f (fst x2) == y)
--
-- $> twoSum [1,2,3,4,10,11,12,2,3] 5
